python/lsst/ts/scheduler/model.py

```diff
@@ -0,0 +1,995 @@
+from lsst.ts.astrosky.model import AstronomicalSkyModel
+from lsst.ts.dateloc import ObservatoryLocation
+from lsst.ts.idl.enums import Script
+from lsst.ts.observatory.model import ObservatoryModel, ObservatoryState
+from rubin_sim.site_models.cloudModel import CloudModel
+from rubin_sim.site_models.downtimeModel import DowntimeModel
+from rubin_sim.site_models.seeingModel import SeeingModel
+
+from lsst.ts import utils
+
+from .driver import Driver
+from .driver.driver_target import DriverTarget
+from .driver.survey_topology import SurveyTopology
+from .telemetry_stream_handler import TelemetryStreamHandler
+from .utils.csc_utils import NonFinalStates, is_uri, is_valid_efd_query
+from .utils.exceptions import UpdateTelemetryError
+from .utils.scheduled_targets_info import ScheduledTargetsInfo
+
+
+class Model:
+    """Scheduler model class.
+
+    This class implements the business logic of the Scheduler. The CSC
+    instantiate the model and offload the operations to it, which allows for
+    better separation of concerns and cleaner code on the CSC part.
+
+    Attributes
+    ----------
+    telemetry_stream_handler : `TelemetryStreamHandler`
+        Object to handle telemetry stream.
+    models : `dict`[`str`, `Any`]
+        Dictionary to store the scheduler models.
+    raw_telemetry : `dict`[`str`, `Any`]
+        Dictionary to store raw telemetry.
+    driver : `Driver`
+        Lock for the scheduler detailed state.
+    """
+
+    def __init__(self, log) -> None:
+
+        self.log = log.getChild(type(self).__name__)
+
+        self.telemetry_stream_handler = None
+
+        # How far to step into the future when there's not targets in seconds
+        self.time_delta_no_target = 30.0
+
+        # Dictionary to store the scheduler models
+        self.models = dict()
+
+        # Dictionary to store raw telemetry
+        self.raw_telemetry = dict()
+
+        # Dictionaries to store information about the scripts put on the queue
+        self.script_info = dict()
+
+        # Scheduler driver instance.
+        self.driver = None
+
+        self.max_scripts = 0
+
+        self.startup_types = dict(
+            HOT=self.configure_driver_hot,
+            WARM=self.configure_driver_warm,
+            COLD=self.configure_driver_cold,
+        )
+
+    async def configure(self, config):
+
+        self.log.debug("Configuring telemetry streams.")
+
+        self.max_scripts = config.max_scripts
+
+        if len(self.raw_telemetry) == 0:
+            self.log.warning("Telemetry stream not initialized. Initializing...")
+            self.init_telemetry()
+
+        await self.configure_telemetry_streams(config.telemetry)
+
+        self.log.debug("Configuring models.")
+
+        if len(self.models) == 0:
+            self.log.warning("Models are not initialized. Initializing...")
+            self.init_models()
+
+        for model in self.models:
+            # TODO: This check will give us time to implement the required
```

Comment:
All TODOs require a Jira ticket so please add the ticket number here for future reference.